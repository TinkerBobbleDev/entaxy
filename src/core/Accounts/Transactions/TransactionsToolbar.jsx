import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'
import IconButton from '@material-ui/core/IconButton'
import Tooltip from '@material-ui/core/Tooltip'
import AddIcon from '@material-ui/icons/Add'
import Icon from '@mdi/react'
import DeleteIcon from '@material-ui/icons/Delete'
import { mdiImport } from '@mdi/js'
import InputBase from '@material-ui/core/InputBase'
import InputAdornment from '@material-ui/core/InputAdornment'
import SearchIcon from '@material-ui/icons/Search'
import { fade } from '@material-ui/core/styles/colorManipulator'
import confirm from '../../../util/confirm'
import TableToolbar from '../../../common/TableToolbar'
import LinkTo from '../../../common/LinkTo'

const styles = (theme) => ({
  importButton: {
    fill: theme.palette.text.secondary
  },
  buttons: {
    display: 'flex'
  },
  search: {
    position: 'relative',
    borderRadius: theme.shape.borderRadius * 2,
    backgroundColor: fade(theme.palette.grey[400], 0.15),
    marginRight: theme.spacing(2),
    marginLeft: 0,
    width: '100%',
    '&:hover': {
      backgroundColor: fade(theme.palette.grey[400], 0.25)
    }
  },
  inputRoot: {
    color: 'inherit',
    width: '100%',
    padding: theme.spacing(2),
    paddingTop: theme.spacing(1),
    paddingBottom: theme.spacing(1)

  },
  inputInput: {
    transition: theme.transitions.create('width'),
    width: '100%',
    [theme.breakpoints.up('md')]: {
      minWidth: 240
    }
  }
})

export class TransactionsToolbarComponent extends React.Component {
  onDelete = () => {
    confirm('Delete selected transactions?', 'Are you sure?').then(() => {
      this.props.handleDelete(this.props.selectedTransactions)
      this.props.resetSelection()
    })
  }

  onChangeSearch = ({ target: { value } }) => {
    const { filterProps } = this.props
    if (value !== '') {
      filterProps.setFilter({
        attr: 'description',
        value: value.toLowerCase()
      })
    } else {
      filterProps.unsetFilter({ attr: 'description' })
    }
  }

  render() {
    const {
      classes,
      account,
      handleNew,
      selectedTransactions,
      filterProps
    } = this.props

    return (
      <TableToolbar
        title={`${account.institution} - ${account.name}`}
        subTitle={account.currency}
        selectedItems={selectedTransactions}
      >
        {selectedTransactions.length > 0 ? (
          <Tooltip title="Delete">
            <IconButton aria-label="Delete" onClick={this.onDelete}>
              <DeleteIcon />
            </IconButton>
          </Tooltip>
        ) : (
          <div className={classes.buttons}>
            <div className={classes.search}>
              <InputBase
                type="search"
                placeholder="Search transactions"
                onChange={this.onChangeSearch}
                value={filterProps.filters.description || ''}
                classes={{
                  root: classes.inputRoot,
                  input: classes.inputInput
                }}
                startAdornment={<InputAdornment position="start"><SearchIcon /></InputAdornment>}
              />
            </div>
            <Tooltip title="Import transaction">
              <IconButton
                aria-label="Import transaction"
                component={LinkTo(`/accounts/${account.id}/import`)}
              >
                <Icon
                  path={mdiImport}
                  size={1}
                  className={classes.importButton}
                />
              </IconButton>
            </Tooltip>
            <Tooltip title="New transaction">
              <IconButton aria-label="New transaction" onClick={handleNew}>
                <AddIcon />
              </IconButton>
            </Tooltip>
          </div>
        )}
      </TableToolbar>
    )
  }
}

TransactionsToolbarComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  account: PropTypes.object.isRequired,
  handleNew: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
  selectedTransactions: PropTypes.array.isRequired,
  resetSelection: PropTypes.func.isRequired,
  filterProps: PropTypes.object.isRequired
}

export default withStyles(styles)(withRouter(TransactionsToolbarComponent))
