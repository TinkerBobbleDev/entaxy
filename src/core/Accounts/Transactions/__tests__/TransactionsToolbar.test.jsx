import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { TransactionsToolbarComponent } from '../TransactionsToolbar'

describe('TransactionsToolbar', () => {
  const account = { id: 1, name: 'Checking', institution: 'TD' }
  const mochHandleNew = jest.fn()
  const mochHandleDelete = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('matches snapshot with no transactions', () => {
    const component = renderer.create((
      <TransactionsToolbarComponent
        classes={{}}
        history={{}}
        account={account}
        handleNew={mochHandleNew}
        handleDelete={mochHandleDelete}
        selectedTransactions={[1, 2, 3, 4]}
      />
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('Component methods', () => {
    const mochHistoryPush = jest.fn()
    const wrapper = shallow((
      <TransactionsToolbarComponent
        classes={{}}
        history={{ push: mochHistoryPush }}
        account={account}
        handleNew={mochHandleNew}
        handleDelete={mochHandleDelete}
        selectedTransactions={[1, 2, 3, 4]}
      />
    ))
    const instance = wrapper.instance()

    it('should return the pageTitle', () => {
      expect(instance.pageTitle(account)).toEqual(`${account.institution} - ${account.name}`)
      expect(instance.pageTitle(null)).toBeNull()
    })

    it('should show the popup menu', () => {
      expect(wrapper.state('anchorEl')).toBeNull()
      expect(wrapper.state('openPopup')).toBe(false)
      instance.showPopup({ currentTarget: <div>something</div> })
      expect(wrapper.state('anchorEl')).toEqual(<div>something</div>)
      expect(wrapper.state('openPopup')).toBe(true)
    })

    it('should close the popup menu', () => {
      expect(wrapper.state('openPopup')).toBe(true)
      instance.closePopup()
      expect(wrapper.state('openPopup')).toBe(false)
    })

    it('should handle the popup selection', () => {
      const importType = 'CSV'
      instance.handlePopupSelection(importType)
      expect(mochHistoryPush).toHaveBeenCalledWith(`/accounts/${account.id}/import/${importType}`)
    })

    it('should call reset selection', () => {
      instance.resetSelection()
      expect(wrapper.state('anchorEl')).toBeNull()
      expect(wrapper.state('openPopup')).toBe(false)
    })
  })
})
