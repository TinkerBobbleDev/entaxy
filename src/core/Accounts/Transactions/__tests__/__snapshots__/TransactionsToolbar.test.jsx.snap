// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsToolbar matches snapshot with no transactions 1`] = `
<div
  className="MuiToolbar-root-7 MuiToolbar-regular-9 MuiToolbar-gutters-8 TableToolbar-root-1 TableToolbar-highlight-2"
>
  <div
    className="TableToolbar-title-6"
  >
    <h6
      className="MuiTypography-root-11 MuiTypography-subtitle1-29 MuiTypography-colorInherit-40"
    >
      4
       selected
    </h6>
  </div>
  <div
    className="TableToolbar-spacer-3"
  />
  <div
    className="TableToolbar-actions-4"
  >
    <button
      aria-describedby={null}
      aria-label="Delete"
      className="MuiButtonBase-root-60 MuiIconButton-root-54"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      title="Delete"
      type="button"
    >
      <span
        className="MuiIconButton-label-59"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-63"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-72"
      />
    </button>
  </div>
</div>
`;
